modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryMemoryTracker.java

```diff
@@ -203,74 +235,239 @@ public boolean isOffloadingEnabled() {
+        checkClosed();
 
-    /**
-     * @return Total bytes reserved by current query.
-     */
-    public long totalReserved() {
-        return totalReserved;
-    }
+        H2MemoryTracker child = new ChildMemoryTracker(this);
 
-    /**
-     * @return Total files number created by current query.
-     */
-    public int filesCreated() {
-        return filesCreated;
-    }
+        children.add(child);
 
-    /** {@inheritDoc} */
-    @Override public synchronized void incrementFilesCreated() {
-        this.filesCreated++;
+        return child;
     }
 
-    /**
-     * @return Query descriptor.
-     */
-    public String queryDescriptor() {
-        return qryDesc;
+    /** {@inheritDoc} */
+    @Override public synchronized void onChildClosed(H2MemoryTracker child) {
+        if (state != CLOSING_OR_CLOSED_STATE)
+            children.remove(child);
     }
 
     /** {@inheritDoc} */
     @Override public String toString() {
         return S.toString(QueryMemoryTracker.class, this);
     }
-}
\ No newline at end of file
+
+    /** */
+    private static class ChildMemoryTracker implements H2MemoryTracker {
+        /** State updater. */
+        private static final AtomicIntegerFieldUpdater<ChildMemoryTracker> STATE_UPDATER
+            = AtomicIntegerFieldUpdater.newUpdater(ChildMemoryTracker.class, "state");
+
+        /** */
+        private final H2MemoryTracker parent;
+
+        /** */
+        private long reserved;
+
+        /** */
+        private long writtenOnDisk;
+
+        /** */
+        private long totalWrittenOnDisk;
+
+        /** */
+        private volatile int state;
+
+        /**
+         * @param parent Parent.
+         */
+        public ChildMemoryTracker(H2MemoryTracker parent) {
+            this.parent = parent;
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean reserve(long size) {
+            checkClosed();
+
+            boolean res;
+            try {
+                res = parent.reserve(size);
+            }
+            catch (IgniteSQLException ex) {
+                reserved += size;
+
+                throw ex;
+            }
+
+            reserved += size;
+
+            return res;
+        }
+
+        /** {@inheritDoc} */
+        @Override public void release(long size) {
+            checkClosed();
+
+            reserved -= size;
+
+            parent.release(size);
+        }
+
+        /** {@inheritDoc} */
+        @Override public long writtenOnDisk() {
+            return writtenOnDisk;
+        }
+
+        /** {@inheritDoc} */
+        @Override public long totalWrittenOnDisk() {
+            return totalWrittenOnDisk;
+        }
+
+        /** {@inheritDoc} */
+        @Override public long reserved() {
+            return reserved;
+        }
+
+        /** {@inheritDoc} */
+        @Override public void swap(long size) {
+            checkClosed();
+
+            parent.swap(size);
```

Comment:
shouldn't `try .. catch` be added here too?

Reply:
no, because we expect exceptions only when you try to reserve memory and quota exceeded and  offloading is disabled